design/026-requirements_traits.md

```diff
@@ -0,0 +1,125 @@
+It seems the potential combinations of dependencies usage is too high to consider approaches like the current one, defining high level names as ``build_requires``, ``test_requires``, ``requires_shared``, etc. does not scale.
+
+
+## Detailed Design
+
+The proposed traits are arguments to the ``requires``, like:
+
+```python
+def requirements(self):
+    self.requires(ref, headers=True, libs=True, run=None, visible=True, …)
+```
+
+There are defaults defined, so the same previous Conan 1.X syntax of class attribute ``requires ="ref"`` and ``self.requires(ref)`` is still possible.
+
+Reminder, following the GNU standard naming:
+
+- Build context: The machine in which the current Conan process and build is running. If we are cross-compiling in a Windows box for an app that will run in a Linux RaspberryPI environment, the build context is the Windows machine.
+- Host context: The machine that will run the apps and libraries being built, in this example the host context is the Linux running in the RaspberryPI
+
+
+The traits are:
+
+- ``headers`` (default True): This trait indicates that there are headers that are going to be #included from this package at compile time. The dependency will be in the host context.
+- ``libs`` (default True): The dependency contains some library or artifact that will be used at link time of the consumer. The dependency will be in the host context. This trait will be true for direct shared and static libraries, but could be false for indirect static libraries that are consumed via a shared library.
+- ``build`` (default False): This dependency is a build tool, an application or executable, like cmake, that is used exclusively at build time, it is not linked/embedded into binaries, and will be in the build context.
+- ``run`` (default None): This dependency contains some executables, either apps or shared libraries that need to be available to execute (typically in the path, or other system env-vars). By default is None, indicating that Conan can try to deduce if the dependency is needed to execute (if ``options.shared`` is True). This trait can be True for ``build=False``, in that case, the package will contain some executables that can run in the host system when installing it, typically like an end-user application. This trait can be True for ``build=True``, the package will contain executables that will run in the build context, typically while being used to build other packages.
+- ``visible`` (default True): This ``require`` will be propagated downstream, even if it doesn’t propagate ``headers``, ``libs`` or ``run`` traits. Requirements that propagate downstream can cause version conflicts. This is by default True, because in most cases, having 2 different versions of the same library in the same dependency graph is at least complicated, if not directly violating ODR or causing linking errors.
```

Comment:
> Do you envision self.requires(... ,headers=True, libs=True, visible=False) as a valid combination?

Yes, this would be a completely private requirement, isolated, non-conflicting library. Might happen in applications or shared libraries linking static or header libraries. By default ``visible=True``, because it seems a sensible default to assume that the same version of the transitive dependency is desired, even if you have multiple shared libraries and applications in the same dependency graph. Putting the ``visible=False`` will allow to have shared1->zlib1 and shared2->zlib2 in the same dependency graph (app -> (shared1 and shared2).

> Is a requires(visible=False) resolved in the host profile but not propagated (like requires(private=True was), or is the recipe not loaded at all (like build_requires)?

As by the previous accepted proposal https://github.com/conan-io/tribe/pull/20, the graph will always be complete in terms of recipes, including build_requires, and including visible=False requires. The trait will make:
- To not create conflicts down the graph.
- To not propagate the headers and linkage requirements further down the graph
- If the final binary is not necessary (existing downstream binary, no need to rebuild from source), then fetching the package binary might be avoided, for performance.

> How do force and visible interact? I.e. does visible=False make this requirement ignore a force (since the downstream can't see that it even exists) or is it still subject to being overridden (even though they don't see the result)?

The ``force`` trait defines a direct dependency to the caller. But it will not force it to "invisible" upstream requirements, as they are not propagated, and they do not conflict. The ``force`` trait does solve conflicts forcing the version, but do not forces things that do not exist for it.

The ``override`` is similar, it can override dependencies, but only the visible ones.

This is the proposal, because the ``visible=False`` is something to be used mainly when you want to allow different versions of the dependencies to co-exist in the same graph at the same time. It is probably only needed in special circumstances, and many dependencies graph will want to use the same versions upstream. But if the graph defined ``visible=False`` to allow precisely this thing, because most likely some of the packages have limitations to upgrade versions, then the ``force`` or ``override`` will probably not solve the issue. Because there will be different requires upstream like requiring zlib1, zlib2 and zlib3 by different packages, and if force/override did that, it will override all of them, and some will not work. Trying to attempt to do something like "override only the zlib dependency of the boost dependency" seems a rabbit hole.


Regarding your use case, I think this proposal will help in propagating better the information, headers, linkage, but the other part that you are describing is more related to the binary model, which this proposal doesn't address.

If you create such ABI stable package, even if it isolates completely the implementation details, even the language, the ``package_id`` of such package will probably be a function of its dependencies settings. Compiling your dependencies in a more modern version of Clang, will most likely create better code, with less bugs, and your final binary gathering that native code, should have a different ``package_id`` than the previous one that was compiled using a gcc 4.9. So passing the ``compiler=clang, compiler.version=9`` to the installation of your package is still very relevant to get the right package binary.

The only way to achieve this level of isolation and independence, via a very aggressive configuration/settings erasure, would be the concept of "re-packaging", and introducing a completely new type of relationships that could model this (I don't think this could be done just by a "trait", it seems like a completely different thing, like the ``python_requires``). Given that such model will also introduce a physical barrier for CI flows, and Conan computation of lockfiles, build-order, etc to work properly, this needs to be carefully considered.

Reply:
> As by the previous accepted proposal #20, the graph will always be complete in terms of recipes, including build_requires, and including visible=False requires. 

Yes, we discussed the problems there too, up through https://github.com/conan-io/tribe/pull/20#issuecomment-799796255, but it didn't really come to a conclusion (which was OK at that time, but this is the next piece to get detailed). So... when you say "complete in terms of recipes", does really mean just the recipe, or do you meant you want to resolve the package_id/package_info too (even when it won't get downloaded/propagated downstream). The former seems OK, the latter quite problematic.

Or to make it specific, if I have:

1. `App` with requires=`FooLib`, 
2. FooLib recipe with self.requires("BarLib",visible=False), 
    FooLib has a well-controlled ABI, and a correspondingly generous compatible_packages allowing the selected package_id to be quite different than App's profile
3. BarLib recipe with either settings restrictions or a configure() method might throw `ConanInvalidConfiguration`
4. The host context of App is not suitable to build or directly call BarLib (violates those settings restriction or configure() would fail).

Will `visible=False` stop after resolving/loading the BarLib recipe but before propagating settings to it (since conan doesn't need to select a binary for an invisible package)? Or will App's profile settings be propagated to Bar despite the fact binaries are not needed, triggering the ConanInvalidConfiguration and making the `conan install`? It feels very strange to require that App use a custom profile picking per-package settings for BarLib (and all its transitive requirements), when App isn't even going to be told that BarLib exists in the end.

Obviously there must have been some other profile used as the host context while building FooLib (corresponding to the binary chosen by compatible_package) which *was* compatible with BarLib. But App has no direct means to know what that profile was. The conaninfo.txt for the selected FooLib binary could know (there must have been something that worked when it was built), but compatible_packages are defined not to influence the dependency graph, and are in any case processed too late to help.

Also, since it's not only requirements but even build_requirements that will now have this behavior, similar questions arise regarding the build context too. If FooLib has an existing binary with suitable ABI, but a build_requirement that is not compatible with App's build context (perhaps App is being cross-compiled but FooLib doesn't support that). Such cases where different packages need different build environments cannot support `conan install App --build FooLib`, but they're possibly today and IMO should continue to be possible when the necessary binaries have already been created and are available in the cache or on artifactory.

For the other two questions, you gave the answer I was hoping for :-)

Reply:
> The only way to achieve this level of isolation and independence, via a very aggressive configuration/settings erasure, would be the concept of "re-packaging", 

Yes,that's basically what we do now - there are commonly two recipes, one for obtaining dependencies and building the binaries, and an entirely different one (to be used used with export-pkg) to copy things into the cache for upload consumption or upload. The latter simply fails to mention the majority of the settings/requirements (omitting all those which would be `visible=False` and are not important to a consumer). This course has lots of drawbacks: there's duplication that one can mess up accidentally not synchronize when changing the two recipes, one can never use `conan create` or `--build missing`, and the conaninfo.txt entirely fails to record what requirements/settings the uploaded binary really is (even in the `[full_requires]/`[full_settings]` that would normally know about things that were excluded from package_id). 

Hence I'd love to see things evolve to where `visible=False` and `compatible_packages` become adequate for this kind of library :-)